// @y/y/dist/index-R7GxO-36.js
'use strict';

var Skip = require('./Skip-j0kX7pdq.js');
require('lib0/function');
require('lib0/decoding');
require('lib0/binary');
require('lib0/map');

/** eslint-env browser */


const glo = /** @type {any} */ (typeof globalThis !== 'undefined'
  ? globalThis
  : typeof window !== 'undefined'
    ? window
    // @ts-ignore
    : typeof global !== 'undefined' ? global : {});

const importIdentifier = '__ $YJS$ __';

if (glo[importIdentifier] === true) {
  /**
   * Dear reader of this message. Please take this seriously.
   *
   * If you see this message, make sure that you only import one version of Yjs. In many cases,
   * your package manager installs two versions of Yjs that are used by different packages within your project.
   * Another reason for this message is that some parts of your project use the commonjs version of Yjs
   * and others use the EcmaScript version of Yjs.
   *
   * This often leads to issues that are hard to debug. We often need to perform constructor checks,
   * e.g. `struct instanceof GC`. If you imported different versions of Yjs, it is impossible for us to
   * do the constructor checks anymore - which might break the CRDT algorithm.
   *
   * https://github.com/yjs/yjs/issues/438
   */
  console.error('Yjs was already imported. This breaks constructor checks and will lead to issues! - https://github.com/yjs/yjs/issues/438');
}
glo[importIdentifier] = true;

var Y = /*#__PURE__*/Object.freeze({
  __proto__: null,
  AbsolutePosition: Skip.AbsolutePosition,
  AbstractAttributionManager: Skip.AbstractAttributionManager,
  AbstractConnector: Skip.AbstractConnector,
  AbstractStruct: Skip.AbstractStruct,
  AbstractType: Skip.AbstractType,
  Array: Skip.YArray,
  Attribution: Skip.AttributionItem,
  ContentAny: Skip.ContentAny,
  ContentBinary: Skip.ContentBinary,
  ContentDeleted: Skip.ContentDeleted,
  ContentDoc: Skip.ContentDoc,
  ContentEmbed: Skip.ContentEmbed,
  ContentFormat: Skip.ContentFormat,
  ContentJSON: Skip.ContentJSON,
  ContentString: Skip.ContentString,
  ContentType: Skip.ContentType,
  DiffAttributionManager: Skip.DiffAttributionManager,
  Doc: Skip.Doc,
  GC: Skip.GC,
  ID: Skip.ID,
  IdMap: Skip.IdMap,
  IdSet: Skip.IdSet,
  Item: Skip.Item,
  Map: Skip.YMap,
  RelativePosition: Skip.RelativePosition,
  Skip: Skip.Skip,
  Snapshot: Skip.Snapshot,
  Text: Skip.YText,
  Transaction: Skip.Transaction,
  TwosetAttributionManager: Skip.TwosetAttributionManager,
  UndoManager: Skip.UndoManager,
  UpdateDecoderV1: Skip.UpdateDecoderV1,
  UpdateDecoderV2: Skip.UpdateDecoderV2,
  UpdateEncoderV1: Skip.UpdateEncoderV1,
  UpdateEncoderV2: Skip.UpdateEncoderV2,
  XmlElement: Skip.YXmlElement,
  XmlFragment: Skip.YXmlFragment,
  XmlHook: Skip.YXmlHook,
  XmlText: Skip.YXmlText,
  YEvent: Skip.YEvent,
  applyUpdate: Skip.applyUpdate,
  applyUpdateV2: Skip.applyUpdateV2,
  cleanupYTextFormatting: Skip.cleanupYTextFormatting,
  cloneDoc: Skip.cloneDoc,
  compareIDs: Skip.compareIDs,
  compareRelativePositions: Skip.compareRelativePositions,
  convertUpdateFormatV1ToV2: Skip.convertUpdateFormatV1ToV2,
  convertUpdateFormatV2ToV1: Skip.convertUpdateFormatV2ToV1,
  createAbsolutePositionFromRelativePosition: Skip.createAbsolutePositionFromRelativePosition,
  createAttributionItem: Skip.createAttributionItem,
  createAttributionManagerFromDiff: Skip.createAttributionManagerFromDiff,
  createDeleteSetFromStructStore: Skip.createDeleteSetFromStructStore,
  createDocFromSnapshot: Skip.createDocFromSnapshot,
  createID: Skip.createID,
  createIdMap: Skip.createIdMap,
  createIdMapFromIdSet: Skip.createIdMapFromIdSet,
  createIdSet: Skip.createIdSet,
  createInsertionSetFromStructStore: Skip.createInsertSetFromStructStore,
  createRelativePositionFromJSON: Skip.createRelativePositionFromJSON,
  createRelativePositionFromTypeIndex: Skip.createRelativePositionFromTypeIndex,
  createSnapshot: Skip.createSnapshot,
  decodeIdMap: Skip.decodeIdMap,
  decodeRelativePosition: Skip.decodeRelativePosition,
  decodeSnapshot: Skip.decodeSnapshot,
  decodeSnapshotV2: Skip.decodeSnapshotV2,
  decodeStateVector: Skip.decodeStateVector,
  decodeUpdate: Skip.decodeUpdate,
  decodeUpdateV2: Skip.decodeUpdateV2,
  diffDocsToDelta: Skip.diffDocsToDelta,
  diffIdMap: Skip.diffIdMap,
  diffIdSet: Skip.diffIdSet,
  diffUpdate: Skip.diffUpdate,
  diffUpdateV2: Skip.diffUpdateV2,
  emptySnapshot: Skip.emptySnapshot,
  encodeIdMap: Skip.encodeIdMap,
  encodeRelativePosition: Skip.encodeRelativePosition,
  encodeSnapshot: Skip.encodeSnapshot,
  encodeSnapshotV2: Skip.encodeSnapshotV2,
  encodeStateAsUpdate: Skip.encodeStateAsUpdate,
  encodeStateAsUpdateV2: Skip.encodeStateAsUpdateV2,
  encodeStateVector: Skip.encodeStateVector,
  encodeStateVectorFromUpdate: Skip.encodeStateVectorFromUpdate,
  encodeStateVectorFromUpdateV2: Skip.encodeStateVectorFromUpdateV2,
  equalIdSets: Skip.equalIdSets,
  equalSnapshots: Skip.equalSnapshots,
  findIndexSS: Skip.findIndexSS,
  findRootTypeKey: Skip.findRootTypeKey,
  getItem: Skip.getItem,
  getItemCleanEnd: Skip.getItemCleanEnd,
  getItemCleanStart: Skip.getItemCleanStart,
  getState: Skip.getState,
  getTypeChildren: Skip.getTypeChildren,
  insertIntoIdMap: Skip.insertIntoIdMap,
  insertIntoIdSet: Skip.insertIntoIdSet,
  isParentOf: Skip.isParentOf,
  iterateStructsByIdSet: Skip.iterateStructsByIdSet,
  logType: Skip.logType,
  logUpdate: Skip.logUpdate,
  logUpdateV2: Skip.logUpdateV2,
  mergeIdMaps: Skip.mergeIdMaps,
  mergeIdSets: Skip.mergeIdSets,
  mergeUpdates: Skip.mergeUpdates,
  mergeUpdatesV2: Skip.mergeUpdatesV2,
  noAttributionsManager: Skip.noAttributionsManager,
  obfuscateUpdate: Skip.obfuscateUpdate,
  obfuscateUpdateV2: Skip.obfuscateUpdateV2,
  readIdMap: Skip.readIdMap,
  readIdSet: Skip.readIdSet,
  readUpdate: Skip.readUpdate,
  readUpdateIdRanges: Skip.readUpdateIdRanges,
  readUpdateIdRangesV2: Skip.readUpdateIdRangesV2,
  readUpdateV2: Skip.readUpdateV2,
  relativePositionToJSON: Skip.relativePositionToJSON,
  snapshot: Skip.snapshot,
  snapshotContainsUpdate: Skip.snapshotContainsUpdate,
  transact: Skip.transact,
  tryGc: Skip.tryGc,
  typeListToArraySnapshot: Skip.typeListToArraySnapshot,
  typeMapGetAllSnapshot: Skip.typeMapGetAllSnapshot,
  typeMapGetSnapshot: Skip.typeMapGetSnapshot
});

exports.Y = Y;
