// @y/y/dist/index-DyTeTfmj.js
import { ah as AbsolutePosition, bF as AbstractAttributionManager, A as AbstractConnector, cy as AbstractStruct, bU as AbstractType, cf as YArray, bl as AttributionItem, cN as ContentAny, cB as ContentBinary, cD as ContentDeleted, cF as ContentDoc, cH as ContentEmbed, cJ as ContentFormat, cL as ContentJSON, cP as ContentString, cZ as ContentType, bJ as DiffAttributionManager, D as Doc, cA as GC, a6 as ID, bt as IdMap, b as IdSet, d4 as Item, ch as YMap, ae as RelativePosition, d9 as Skip, ar as Snapshot, co as YText, aS as Transaction, bG as TwosetAttributionManager, aZ as UndoManager, U as UpdateDecoderV1, C as UpdateDecoderV2, F as UpdateEncoderV1, H as UpdateEncoderV2, cs as YXmlElement, cq as YXmlFragment, cu as YXmlHook, cw as YXmlText, bh as YEvent, Q as applyUpdate, P as applyUpdateV2, cl as cleanupYTextFormatting, y as cloneDoc, a7 as compareIDs, aq as compareRelativePositions, bf as convertUpdateFormatV1ToV2, bg as convertUpdateFormatV2ToV1, ap as createAbsolutePositionFromRelativePosition, bm as createAttributionItem, bK as createAttributionManagerFromDiff, q as createDeleteSetFromStructStore, aC as createDocFromSnapshot, a8 as createID, by as createIdMap, bs as createIdMapFromIdSet, p as createIdSet, s as createInsertSetFromStructStore, ag as createRelativePositionFromJSON, ak as createRelativePositionFromTypeIndex, ax as createSnapshot, bx as decodeIdMap, ao as decodeRelativePosition, aw as decodeSnapshot, av as decodeSnapshotV2, W as decodeStateVector, b1 as decodeUpdate, b2 as decodeUpdateV2, bN as diffDocsToDelta, bA as diffIdMap, j as diffIdSet, bb as diffUpdate, ba as diffUpdateV2, ay as emptySnapshot, bv as encodeIdMap, am as encodeRelativePosition, au as encodeSnapshot, at as encodeSnapshotV2, T as encodeStateAsUpdate, S as encodeStateAsUpdateV2, $ as encodeStateVector, b6 as encodeStateVectorFromUpdate, b5 as encodeStateVectorFromUpdateV2, v as equalIdSets, as as equalSnapshots, aK as findIndexSS, ab as findRootTypeKey, aM as getItem, aP as getItemCleanEnd, aO as getItemCleanStart, aH as getState, bS as getTypeChildren, bz as insertIntoIdMap, g as insertIntoIdSet, ac as isParentOf, i as iterateStructsByIdSet, ad as logType, a$ as logUpdate, b0 as logUpdateV2, br as mergeIdMaps, m as mergeIdSets, b4 as mergeUpdates, b9 as mergeUpdatesV2, bI as noAttributionsManager, bd as obfuscateUpdate, be as obfuscateUpdateV2, bw as readIdMap, t as readIdSet, O as readUpdate, b8 as readUpdateIdRanges, b7 as readUpdateIdRangesV2, N as readUpdateV2, af as relativePositionToJSON, az as snapshot, aE as snapshotContainsUpdate, aX as transact, aW as tryGc, bY as typeListToArraySnapshot, cd as typeMapGetAllSnapshot, cc as typeMapGetSnapshot } from './Skip-wRT7BKFP.js';
import 'lib0/function';
import 'lib0/decoding';
import 'lib0/binary';
import 'lib0/map';

/** eslint-env browser */


const glo = /** @type {any} */ (typeof globalThis !== 'undefined'
  ? globalThis
  : typeof window !== 'undefined'
    ? window
    // @ts-ignore
    : typeof global !== 'undefined' ? global : {});

const importIdentifier = '__ $YJS$ __';

if (glo[importIdentifier] === true) {
  /**
   * Dear reader of this message. Please take this seriously.
   *
   * If you see this message, make sure that you only import one version of Yjs. In many cases,
   * your package manager installs two versions of Yjs that are used by different packages within your project.
   * Another reason for this message is that some parts of your project use the commonjs version of Yjs
   * and others use the EcmaScript version of Yjs.
   *
   * This often leads to issues that are hard to debug. We often need to perform constructor checks,
   * e.g. `struct instanceof GC`. If you imported different versions of Yjs, it is impossible for us to
   * do the constructor checks anymore - which might break the CRDT algorithm.
   *
   * https://github.com/yjs/yjs/issues/438
   */
  console.error('Yjs was already imported. This breaks constructor checks and will lead to issues! - https://github.com/yjs/yjs/issues/438');
}
glo[importIdentifier] = true;

var Y = /*#__PURE__*/Object.freeze({
  __proto__: null,
  AbsolutePosition: AbsolutePosition,
  AbstractAttributionManager: AbstractAttributionManager,
  AbstractConnector: AbstractConnector,
  AbstractStruct: AbstractStruct,
  AbstractType: AbstractType,
  Array: YArray,
  Attribution: AttributionItem,
  ContentAny: ContentAny,
  ContentBinary: ContentBinary,
  ContentDeleted: ContentDeleted,
  ContentDoc: ContentDoc,
  ContentEmbed: ContentEmbed,
  ContentFormat: ContentFormat,
  ContentJSON: ContentJSON,
  ContentString: ContentString,
  ContentType: ContentType,
  DiffAttributionManager: DiffAttributionManager,
  Doc: Doc,
  GC: GC,
  ID: ID,
  IdMap: IdMap,
  IdSet: IdSet,
  Item: Item,
  Map: YMap,
  RelativePosition: RelativePosition,
  Skip: Skip,
  Snapshot: Snapshot,
  Text: YText,
  Transaction: Transaction,
  TwosetAttributionManager: TwosetAttributionManager,
  UndoManager: UndoManager,
  UpdateDecoderV1: UpdateDecoderV1,
  UpdateDecoderV2: UpdateDecoderV2,
  UpdateEncoderV1: UpdateEncoderV1,
  UpdateEncoderV2: UpdateEncoderV2,
  XmlElement: YXmlElement,
  XmlFragment: YXmlFragment,
  XmlHook: YXmlHook,
  XmlText: YXmlText,
  YEvent: YEvent,
  applyUpdate: applyUpdate,
  applyUpdateV2: applyUpdateV2,
  cleanupYTextFormatting: cleanupYTextFormatting,
  cloneDoc: cloneDoc,
  compareIDs: compareIDs,
  compareRelativePositions: compareRelativePositions,
  convertUpdateFormatV1ToV2: convertUpdateFormatV1ToV2,
  convertUpdateFormatV2ToV1: convertUpdateFormatV2ToV1,
  createAbsolutePositionFromRelativePosition: createAbsolutePositionFromRelativePosition,
  createAttributionItem: createAttributionItem,
  createAttributionManagerFromDiff: createAttributionManagerFromDiff,
  createDeleteSetFromStructStore: createDeleteSetFromStructStore,
  createDocFromSnapshot: createDocFromSnapshot,
  createID: createID,
  createIdMap: createIdMap,
  createIdMapFromIdSet: createIdMapFromIdSet,
  createIdSet: createIdSet,
  createInsertionSetFromStructStore: createInsertSetFromStructStore,
  createRelativePositionFromJSON: createRelativePositionFromJSON,
  createRelativePositionFromTypeIndex: createRelativePositionFromTypeIndex,
  createSnapshot: createSnapshot,
  decodeIdMap: decodeIdMap,
  decodeRelativePosition: decodeRelativePosition,
  decodeSnapshot: decodeSnapshot,
  decodeSnapshotV2: decodeSnapshotV2,
  decodeStateVector: decodeStateVector,
  decodeUpdate: decodeUpdate,
  decodeUpdateV2: decodeUpdateV2,
  diffDocsToDelta: diffDocsToDelta,
  diffIdMap: diffIdMap,
  diffIdSet: diffIdSet,
  diffUpdate: diffUpdate,
  diffUpdateV2: diffUpdateV2,
  emptySnapshot: emptySnapshot,
  encodeIdMap: encodeIdMap,
  encodeRelativePosition: encodeRelativePosition,
  encodeSnapshot: encodeSnapshot,
  encodeSnapshotV2: encodeSnapshotV2,
  encodeStateAsUpdate: encodeStateAsUpdate,
  encodeStateAsUpdateV2: encodeStateAsUpdateV2,
  encodeStateVector: encodeStateVector,
  encodeStateVectorFromUpdate: encodeStateVectorFromUpdate,
  encodeStateVectorFromUpdateV2: encodeStateVectorFromUpdateV2,
  equalIdSets: equalIdSets,
  equalSnapshots: equalSnapshots,
  findIndexSS: findIndexSS,
  findRootTypeKey: findRootTypeKey,
  getItem: getItem,
  getItemCleanEnd: getItemCleanEnd,
  getItemCleanStart: getItemCleanStart,
  getState: getState,
  getTypeChildren: getTypeChildren,
  insertIntoIdMap: insertIntoIdMap,
  insertIntoIdSet: insertIntoIdSet,
  isParentOf: isParentOf,
  iterateStructsByIdSet: iterateStructsByIdSet,
  logType: logType,
  logUpdate: logUpdate,
  logUpdateV2: logUpdateV2,
  mergeIdMaps: mergeIdMaps,
  mergeIdSets: mergeIdSets,
  mergeUpdates: mergeUpdates,
  mergeUpdatesV2: mergeUpdatesV2,
  noAttributionsManager: noAttributionsManager,
  obfuscateUpdate: obfuscateUpdate,
  obfuscateUpdateV2: obfuscateUpdateV2,
  readIdMap: readIdMap,
  readIdSet: readIdSet,
  readUpdate: readUpdate,
  readUpdateIdRanges: readUpdateIdRanges,
  readUpdateIdRangesV2: readUpdateIdRangesV2,
  readUpdateV2: readUpdateV2,
  relativePositionToJSON: relativePositionToJSON,
  snapshot: snapshot,
  snapshotContainsUpdate: snapshotContainsUpdate,
  transact: transact,
  tryGc: tryGc,
  typeListToArraySnapshot: typeListToArraySnapshot,
  typeMapGetAllSnapshot: typeMapGetAllSnapshot,
  typeMapGetSnapshot: typeMapGetSnapshot
});

export { Y };
