// @y/y/dist/internals.cjs
'use strict';

var Skip = require('./Skip-j0kX7pdq.js');
require('lib0/observable');
require('lib0/array');
require('lib0/math');
require('lib0/encoding');
require('lib0/decoding');
require('lib0/traits');
require('lib0/random');
require('lib0/map');
require('lib0/promise');
require('lib0/buffer');
require('lib0/error');
require('lib0/function');
require('lib0/set');
require('lib0/logging');
require('lib0/time');
require('lib0/binary');
require('lib0/string');
require('lib0/delta');
require('lib0/hash/rabin');
require('lib0/iterator');
require('lib0/object');
require('lib0/environment');
require('lib0/schema');



exports.AbsolutePosition = Skip.AbsolutePosition;
exports.AbstractAttributionManager = Skip.AbstractAttributionManager;
exports.AbstractConnector = Skip.AbstractConnector;
exports.AbstractContent = Skip.AbstractContent;
exports.AbstractStruct = Skip.AbstractStruct;
exports.AbstractType = Skip.AbstractType;
exports.ArraySearchMarker = Skip.ArraySearchMarker;
exports.AttrRange = Skip.AttrRange;
exports.AttrRanges = Skip.AttrRanges;
exports.AttributedContent = Skip.AttributedContent;
exports.AttributionItem = Skip.AttributionItem;
exports.ContentAny = Skip.ContentAny;
exports.ContentBinary = Skip.ContentBinary;
exports.ContentDeleted = Skip.ContentDeleted;
exports.ContentDoc = Skip.ContentDoc;
exports.ContentEmbed = Skip.ContentEmbed;
exports.ContentFormat = Skip.ContentFormat;
exports.ContentJSON = Skip.ContentJSON;
exports.ContentString = Skip.ContentString;
exports.ContentType = Skip.ContentType;
exports.DSDecoderV1 = Skip.DSDecoderV1;
exports.DSDecoderV2 = Skip.DSDecoderV2;
exports.DiffAttributionManager = Skip.DiffAttributionManager;
exports.Doc = Skip.Doc;
exports.EventHandler = Skip.EventHandler;
exports.GC = Skip.GC;
exports.ID = Skip.ID;
exports.IdMap = Skip.IdMap;
exports.IdRange = Skip.IdRange;
exports.IdRanges = Skip.IdRanges;
exports.IdSet = Skip.IdSet;
exports.IdSetEncoderV1 = Skip.IdSetEncoderV1;
exports.IdSetEncoderV2 = Skip.IdSetEncoderV2;
exports.Item = Skip.Item;
exports.ItemTextListPosition = Skip.ItemTextListPosition;
exports.LazyStructReader = Skip.LazyStructReader;
exports.LazyStructWriter = Skip.LazyStructWriter;
exports.MaybeIdRange = Skip.MaybeIdRange;
exports.NoAttributionsManager = Skip.NoAttributionsManager;
exports.RelativePosition = Skip.RelativePosition;
exports.Skip = Skip.Skip;
exports.Snapshot = Skip.Snapshot;
exports.SnapshotAttributionManager = Skip.SnapshotAttributionManager;
exports.StackItem = Skip.StackItem;
exports.StructSet = Skip.StructSet;
exports.StructStore = Skip.StructStore;
exports.Transaction = Skip.Transaction;
exports.TwosetAttributionManager = Skip.TwosetAttributionManager;
exports.UndoManager = Skip.UndoManager;
exports.UpdateDecoderV1 = Skip.UpdateDecoderV1;
exports.UpdateDecoderV2 = Skip.UpdateDecoderV2;
exports.UpdateEncoderV1 = Skip.UpdateEncoderV1;
exports.UpdateEncoderV2 = Skip.UpdateEncoderV2;
exports.YArray = Skip.YArray;
exports.YArrayRefID = Skip.YArrayRefID;
exports.YEvent = Skip.YEvent;
exports.YMap = Skip.YMap;
exports.YMapRefID = Skip.YMapRefID;
exports.YText = Skip.YText;
exports.YTextRefID = Skip.YTextRefID;
exports.YXmlElement = Skip.YXmlElement;
exports.YXmlElementRefID = Skip.YXmlElementRefID;
exports.YXmlFragment = Skip.YXmlFragment;
exports.YXmlFragmentRefID = Skip.YXmlFragmentRefID;
exports.YXmlHook = Skip.YXmlHook;
exports.YXmlHookRefID = Skip.YXmlHookRefID;
exports.YXmlText = Skip.YXmlText;
exports.YXmlTextRefID = Skip.YXmlTextRefID;
exports._createInsertSliceFromStructs = Skip._createInsertSliceFromStructs;
exports._deleteRangeFromIdSet = Skip._deleteRangeFromIdSet;
exports._diffSet = Skip._diffSet;
exports._insertIntoIdSet = Skip._insertIntoIdSet;
exports._intersectSets = Skip._intersectSets;
exports.addChangedTypeToTransaction = Skip.addChangedTypeToTransaction;
exports.addEventHandlerListener = Skip.addEventHandlerListener;
exports.addStruct = Skip.addStruct;
exports.addStructToIdSet = Skip.addStructToIdSet;
exports.addToIdSet = Skip.addToIdSet;
exports.applyUpdate = Skip.applyUpdate;
exports.applyUpdateV2 = Skip.applyUpdateV2;
exports.attributionJsonSchema = Skip.attributionJsonSchema;
exports.callEventHandlerListeners = Skip.callEventHandlerListeners;
exports.callTypeObservers = Skip.callTypeObservers;
exports.cleanupYTextAfterTransaction = Skip.cleanupYTextAfterTransaction;
exports.cleanupYTextFormatting = Skip.cleanupYTextFormatting;
exports.cloneDoc = Skip.cloneDoc;
exports.compareIDs = Skip.compareIDs;
exports.compareRelativePositions = Skip.compareRelativePositions;
exports.contentRefs = Skip.contentRefs;
exports.convertUpdateFormat = Skip.convertUpdateFormat;
exports.convertUpdateFormatV1ToV2 = Skip.convertUpdateFormatV1ToV2;
exports.convertUpdateFormatV2ToV1 = Skip.convertUpdateFormatV2ToV1;
exports.createAbsolutePosition = Skip.createAbsolutePosition;
exports.createAbsolutePositionFromRelativePosition = Skip.createAbsolutePositionFromRelativePosition;
exports.createAttributionFromAttributionItems = Skip.createAttributionFromAttributionItems;
exports.createAttributionItem = Skip.createAttributionItem;
exports.createAttributionManagerFromDiff = Skip.createAttributionManagerFromDiff;
exports.createAttributionManagerFromSnapshots = Skip.createAttributionManagerFromSnapshots;
exports.createDeleteSetFromStructStore = Skip.createDeleteSetFromStructStore;
exports.createDocFromSnapshot = Skip.createDocFromSnapshot;
exports.createEventHandler = Skip.createEventHandler;
exports.createID = Skip.createID;
exports.createIdMap = Skip.createIdMap;
exports.createIdMapFromIdSet = Skip.createIdMapFromIdSet;
exports.createIdSet = Skip.createIdSet;
exports.createInsertSetFromStructStore = Skip.createInsertSetFromStructStore;
exports.createMapIterator = Skip.createMapIterator;
exports.createMaybeAttrRange = Skip.createMaybeAttrRange;
exports.createMaybeIdRange = Skip.createMaybeIdRange;
exports.createRelativePosition = Skip.createRelativePosition;
exports.createRelativePositionFromJSON = Skip.createRelativePositionFromJSON;
exports.createRelativePositionFromTypeIndex = Skip.createRelativePositionFromTypeIndex;
exports.createSnapshot = Skip.createSnapshot;
exports.decodeIdMap = Skip.decodeIdMap;
exports.decodeRelativePosition = Skip.decodeRelativePosition;
exports.decodeSnapshot = Skip.decodeSnapshot;
exports.decodeSnapshotV2 = Skip.decodeSnapshotV2;
exports.decodeStateVector = Skip.decodeStateVector;
exports.decodeUpdate = Skip.decodeUpdate;
exports.decodeUpdateV2 = Skip.decodeUpdateV2;
exports.deleteText = Skip.deleteText;
exports.diffDocsToDelta = Skip.diffDocsToDelta;
exports.diffIdMap = Skip.diffIdMap;
exports.diffIdSet = Skip.diffIdSet;
exports.diffUpdate = Skip.diffUpdate;
exports.diffUpdateV2 = Skip.diffUpdateV2;
exports.emptySnapshot = Skip.emptySnapshot;
exports.encodeIdMap = Skip.encodeIdMap;
exports.encodeRelativePosition = Skip.encodeRelativePosition;
exports.encodeSnapshot = Skip.encodeSnapshot;
exports.encodeSnapshotV2 = Skip.encodeSnapshotV2;
exports.encodeStateAsUpdate = Skip.encodeStateAsUpdate;
exports.encodeStateAsUpdateV2 = Skip.encodeStateAsUpdateV2;
exports.encodeStateVector = Skip.encodeStateVector;
exports.encodeStateVectorFromUpdate = Skip.encodeStateVectorFromUpdate;
exports.encodeStateVectorFromUpdateV2 = Skip.encodeStateVectorFromUpdateV2;
exports.encodeStateVectorV2 = Skip.encodeStateVectorV2;
exports.equalAttrs = Skip.equalAttrs;
exports.equalIdSets = Skip.equalIdSets;
exports.equalSnapshots = Skip.equalSnapshots;
exports.find = Skip.find;
exports.findIndexCleanStart = Skip.findIndexCleanStart;
exports.findIndexInIdRanges = Skip.findIndexInIdRanges;
exports.findIndexSS = Skip.findIndexSS;
exports.findMarker = Skip.findMarker;
exports.findRangeStartInIdRanges = Skip.findRangeStartInIdRanges;
exports.findRootTypeKey = Skip.findRootTypeKey;
exports.followRedone = Skip.followRedone;
exports.generateNewClientId = Skip.generateNewClientId;
exports.getItem = Skip.getItem;
exports.getItemCleanEnd = Skip.getItemCleanEnd;
exports.getItemCleanStart = Skip.getItemCleanStart;
exports.getState = Skip.getState;
exports.getStateVector = Skip.getStateVector;
exports.getTypeChildren = Skip.getTypeChildren;
exports.idmapAttrsEqual = Skip.idmapAttrsEqual;
exports.insertIntoIdMap = Skip.insertIntoIdMap;
exports.insertIntoIdSet = Skip.insertIntoIdSet;
exports.insertText = Skip.insertText;
exports.integrityCheck = Skip.integrityCheck;
exports.intersectMaps = Skip.intersectMaps;
exports.intersectSets = Skip.intersectSets;
exports.isParentOf = Skip.isParentOf;
exports.isVisible = Skip.isVisible;
exports.iterateStructs = Skip.iterateStructs;
exports.iterateStructsByIdSet = Skip.iterateStructsByIdSet;
exports.keepItem = Skip.keepItem;
exports.logType = Skip.logType;
exports.logUpdate = Skip.logUpdate;
exports.logUpdateV2 = Skip.logUpdateV2;
exports.mergeIdMaps = Skip.mergeIdMaps;
exports.mergeIdSets = Skip.mergeIdSets;
exports.mergeUpdates = Skip.mergeUpdates;
exports.mergeUpdatesV2 = Skip.mergeUpdatesV2;
exports.nextID = Skip.nextID;
exports.noAttributionsManager = Skip.noAttributionsManager;
exports.obfuscateUpdate = Skip.obfuscateUpdate;
exports.obfuscateUpdateV2 = Skip.obfuscateUpdateV2;
exports.readAndApplyDeleteSet = Skip.readAndApplyDeleteSet;
exports.readContentAny = Skip.readContentAny;
exports.readContentBinary = Skip.readContentBinary;
exports.readContentDeleted = Skip.readContentDeleted;
exports.readContentDoc = Skip.readContentDoc;
exports.readContentEmbed = Skip.readContentEmbed;
exports.readContentFormat = Skip.readContentFormat;
exports.readContentJSON = Skip.readContentJSON;
exports.readContentString = Skip.readContentString;
exports.readContentType = Skip.readContentType;
exports.readID = Skip.readID;
exports.readIdMap = Skip.readIdMap;
exports.readIdSet = Skip.readIdSet;
exports.readItemContent = Skip.readItemContent;
exports.readRelativePosition = Skip.readRelativePosition;
exports.readStateVector = Skip.readStateVector;
exports.readStructSet = Skip.readStructSet;
exports.readUpdate = Skip.readUpdate;
exports.readUpdateIdRanges = Skip.readUpdateIdRanges;
exports.readUpdateIdRangesV2 = Skip.readUpdateIdRangesV2;
exports.readUpdateV2 = Skip.readUpdateV2;
exports.readYArray = Skip.readYArray;
exports.readYMap = Skip.readYMap;
exports.readYText = Skip.readYText;
exports.readYXmlElement = Skip.readYXmlElement;
exports.readYXmlFragment = Skip.readYXmlFragment;
exports.readYXmlHook = Skip.readYXmlHook;
exports.readYXmlText = Skip.readYXmlText;
exports.redoItem = Skip.redoItem;
exports.relativePositionToJSON = Skip.relativePositionToJSON;
exports.removeAllEventHandlerListeners = Skip.removeAllEventHandlerListeners;
exports.removeEventHandlerListener = Skip.removeEventHandlerListener;
exports.removeRangesFromStructSet = Skip.removeRangesFromStructSet;
exports.replaceStruct = Skip.replaceStruct;
exports.snapshot = Skip.snapshot;
exports.snapshotContainsUpdate = Skip.snapshotContainsUpdate;
exports.snapshotContainsUpdateV2 = Skip.snapshotContainsUpdateV2;
exports.splitItem = Skip.splitItem;
exports.splitSnapshotAffectedStructs = Skip.splitSnapshotAffectedStructs;
exports.splitStruct = Skip.splitStruct;
exports.structGCRefNumber = Skip.structGCRefNumber;
exports.structSkipRefNumber = Skip.structSkipRefNumber;
exports.transact = Skip.transact;
exports.tryGc = Skip.tryGc;
exports.typeListCreateIterator = Skip.typeListCreateIterator;
exports.typeListDelete = Skip.typeListDelete;
exports.typeListForEach = Skip.typeListForEach;
exports.typeListForEachSnapshot = Skip.typeListForEachSnapshot;
exports.typeListGet = Skip.typeListGet;
exports.typeListInsertGenerics = Skip.typeListInsertGenerics;
exports.typeListInsertGenericsAfter = Skip.typeListInsertGenericsAfter;
exports.typeListMap = Skip.typeListMap;
exports.typeListPushGenerics = Skip.typeListPushGenerics;
exports.typeListSlice = Skip.typeListSlice;
exports.typeListToArray = Skip.typeListToArray;
exports.typeListToArraySnapshot = Skip.typeListToArraySnapshot;
exports.typeMapDelete = Skip.typeMapDelete;
exports.typeMapGet = Skip.typeMapGet;
exports.typeMapGetAll = Skip.typeMapGetAll;
exports.typeMapGetAllSnapshot = Skip.typeMapGetAllSnapshot;
exports.typeMapGetDelta = Skip.typeMapGetDelta;
exports.typeMapGetSnapshot = Skip.typeMapGetSnapshot;
exports.typeMapHas = Skip.typeMapHas;
exports.typeMapSet = Skip.typeMapSet;
exports.typeRefs = Skip.typeRefs;
exports.updateMarkerChanges = Skip.updateMarkerChanges;
exports.warnPrematureAccess = Skip.warnPrematureAccess;
exports.writeClientsStructs = Skip.writeClientsStructs;
exports.writeDocumentStateVector = Skip.writeDocumentStateVector;
exports.writeID = Skip.writeID;
exports.writeIdMap = Skip.writeIdMap;
exports.writeIdSet = Skip.writeIdSet;
exports.writeRelativePosition = Skip.writeRelativePosition;
exports.writeStateAsUpdate = Skip.writeStateAsUpdate;
exports.writeStateVector = Skip.writeStateVector;
exports.writeStructsFromIdSet = Skip.writeStructsFromIdSet;
exports.writeStructsFromTransaction = Skip.writeStructsFromTransaction;
exports.writeUpdateMessageFromTransaction = Skip.writeUpdateMessageFromTransaction;
