// @y/y/dist/testHelper.cjs
'use strict';

var t = require('lib0/testing');
var prng = require('lib0/prng');
var encoding = require('lib0/encoding');
var decoding = require('lib0/decoding');
var syncProtocol = require('@y/protocols/sync');
var object = require('lib0/object');
var map = require('lib0/map');
var yjs = require('./index-R7GxO-36.js');
var math = require('lib0/math');
var list = require('lib0/list');
var delta = require('lib0/delta');
var Skip = require('./Skip-j0kX7pdq.js');
require('lib0/function');
require('lib0/binary');
require('lib0/observable');
require('lib0/array');
require('lib0/traits');
require('lib0/random');
require('lib0/promise');
require('lib0/buffer');
require('lib0/error');
require('lib0/set');
require('lib0/logging');
require('lib0/time');
require('lib0/string');
require('lib0/hash/rabin');
require('lib0/iterator');
require('lib0/environment');
require('lib0/schema');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var t__namespace = /*#__PURE__*/_interopNamespaceDefault(t);
var prng__namespace = /*#__PURE__*/_interopNamespaceDefault(prng);
var encoding__namespace = /*#__PURE__*/_interopNamespaceDefault(encoding);
var decoding__namespace = /*#__PURE__*/_interopNamespaceDefault(decoding);
var syncProtocol__namespace = /*#__PURE__*/_interopNamespaceDefault(syncProtocol);
var object__namespace = /*#__PURE__*/_interopNamespaceDefault(object);
var map__namespace = /*#__PURE__*/_interopNamespaceDefault(map);
var math__namespace = /*#__PURE__*/_interopNamespaceDefault(math);
var list__namespace = /*#__PURE__*/_interopNamespaceDefault(list);
var delta__namespace = /*#__PURE__*/_interopNamespaceDefault(delta);

if (typeof window !== 'undefined') {
  // @ts-ignore
  window.Y = yjs.Y; // eslint-disable-line
}

/**
 * @param {TestYInstance} y // publish message created by `y` to all other online clients
 * @param {Uint8Array} m
 */
const broadcastMessage = (y, m) => {
  if (y.tc.onlineConns.has(y)) {
    y.tc.onlineConns.forEach(remoteYInstance => {
      if (remoteYInstance !== y) {
        remoteYInstance._receive(m, y);
      }
    });
  }
};

exports.useV2 = false;

const encV1 = {
  encodeStateAsUpdate: Skip.encodeStateAsUpdate,
  mergeUpdates: Skip.mergeUpdates,
  applyUpdate: Skip.applyUpdate,
  logUpdate: Skip.logUpdate,
  updateEventName: /** @type {'update'} */ ('update'),
  diffUpdate: Skip.diffUpdate
};

const encV2 = {
  encodeStateAsUpdate: Skip.encodeStateAsUpdateV2,
  mergeUpdates: Skip.mergeUpdatesV2,
  applyUpdate: Skip.applyUpdateV2,
  logUpdate: Skip.logUpdateV2,
  updateEventName: /** @type {'updateV2'} */ ('updateV2'),
  diffUpdate: Skip.diffUpdateV2
};

exports.enc = encV1;

const useV1Encoding = () => {
  exports.useV2 = false;
  exports.enc = encV1;
};

const useV2Encoding = () => {
  console.error('sync protocol doesnt support v2 protocol yet, fallback to v1 encoding'); // @Todo
  exports.useV2 = false;
  exports.enc = encV1;
};

class TestYInstance extends Skip.Doc {
  /**
   * @param {TestConnector} testConnector
   * @param {number} clientID
   */
  constructor (testConnector, clientID) {
    super();
    this.userID = clientID; // overwriting clientID
    /**
     * @type {TestConnector}
     */
    this.tc = testConnector;
    /**
     * @type {Map<TestYInstance, Array<Uint8Array>>}
     */
    this.receiving = new Map();
    testConnector.allConns.add(this);
    /**
     * The list of received updates.
     * We are going to merge them later using Y.mergeUpdates and check if the resulting document is correct.
     * @type {Array<Uint8Array<ArrayBuffer>>}
     */
    this.updates = [];
    // set up observe on local model
    this.on(exports.enc.updateEventName, (update, origin) => {
      if (origin !== testConnector) {
        const encoder = encoding__namespace.createEncoder();
        syncProtocol__namespace.writeUpdate(encoder, update);
        broadcastMessage(this, encoding__namespace.toUint8Array(encoder));
      }
      this.updates.push(update);
    });
    this.connect();
  }

  /**
   * Disconnect from TestConnector.
   */
  disconnect () {
    this.receiving = new Map();
    this.tc.onlineConns.delete(this);
  }

  /**
   * Append yourself to the list of known Y instances in testconnector.
   * Also initiate sync with all clients.
   */
  connect () {
    if (!this.tc.onlineConns.has(this)) {
      this.tc.onlineConns.add(this);
      const encoder = encoding__namespace.createEncoder();
      syncProtocol__namespace.writeSyncStep1(encoder, this);
      // publish SyncStep1
      broadcastMessage(this, encoding__namespace.toUint8Array(encoder));
      this.tc.onlineConns.forEach(remoteYInstance => {
        if (remoteYInstance !== this) {
          // remote instance sends instance to this instance
          const encoder = encoding__namespace.createEncoder();
          syncProtocol__namespace.writeSyncStep1(encoder, remoteYInstance);
          this._receive(encoding__namespace.toUint8Array(encoder), remoteYInstance);
        }
      });
    }
  }

  /**
   * Receive a message from another client. This message is only appended to the list of receiving messages.
   * TestConnector decides when this client actually reads this message.
   *
   * @param {Uint8Array} message
   * @param {TestYInstance} remoteClient
   */
  _receive (message, remoteClient) {
    map__namespace.setIfUndefined(this.receiving, remoteClient, () => /** @type {Array<Uint8Array>} */ ([])).push(message);
  }
}

/**
 * Keeps track of TestYInstances.
 *
 * The TestYInstances add/remove themselves from the list of connections maiained in this object.
 * I think it makes sense. Deal with it.
 */
class TestConnector {
  /**
   * @param {prng.PRNG} gen
   */
  constructor (gen) {
    /**
     * @type {Set<TestYInstance>}
     */
    this.allConns = new Set();
    /**
     * @type {Set<TestYInstance>}
     */
    this.onlineConns = new Set();
    /**
     * @type {prng.PRNG}
     */
    this.prng = gen;
  }

  /**
   * Create a new Y instance and add it to the list of connections
   * @param {number} clientID
   */
  createY (clientID) {
    return new TestYInstance(this, clientID)
  }

  /**
   * Choose random connection and flush a random message from a random sender.
   *
   * If this function was unable to flush a message, because there are no more messages to flush, it returns false. true otherwise.
   * @return {boolean}
   */
  flushRandomMessage () {
    const gen = this.prng;
    const conns = Array.from(this.onlineConns).filter(conn => conn.receiving.size > 0);
    if (conns.length > 0) {
      const receiver = prng__namespace.oneOf(gen, conns);
      const [sender, messages] = prng__namespace.oneOf(gen, Array.from(receiver.receiving));
      const m = messages.shift();
      if (messages.length === 0) {
        receiver.receiving.delete(sender);
      }
      if (m === undefined) {
        return this.flushRandomMessage()
      }
      const encoder = encoding__namespace.createEncoder();
      // console.log('receive (' + sender.userID + '->' + receiver.userID + '):\n', syncProtocol.stringifySyncMessage(decoding.createDecoder(m), receiver))
      // do not publish data created when this function is executed (could be ss2 or update message)
      syncProtocol__namespace.readSyncMessage(decoding__namespace.createDecoder(m), encoder, receiver, receiver.tc);
      if (encoding__namespace.length(encoder) > 0) {
        // send reply message
        sender._receive(encoding__namespace.toUint8Array(encoder), receiver);
      }
      return true
    }
    return false
  }

  /**
   * @return {boolean} True iff this function actually flushed something
   */
  flushAllMessages () {
    let didSomething = false;
    while (this.flushRandomMessage()) {
      didSomething = true;
    }
    return didSomething
  }

  reconnectAll () {
    this.allConns.forEach(conn => conn.connect());
  }

  disconnectAll () {
    this.allConns.forEach(conn => conn.disconnect());
  }

  syncAll () {
    this.reconnectAll();
    this.flushAllMessages();
  }

  /**
   * @return {boolean} Whether it was possible to disconnect a random connection.
   */
  disconnectRandom () {
    if (this.onlineConns.size === 0) {
      return false
    }
    prng__namespace.oneOf(this.prng, Array.from(this.onlineConns)).disconnect();
    return true
  }

  /**
   * @return {boolean} Whether it was possible to reconnect a random connection.
   */
  reconnectRandom () {
    /**
     * @type {Array<TestYInstance>}
     */
    const reconnectable = [];
    this.allConns.forEach(conn => {
      if (!this.onlineConns.has(conn)) {
        reconnectable.push(conn);
      }
    });
    if (reconnectable.length === 0) {
      return false
    }
    prng__namespace.oneOf(this.prng, reconnectable).connect();
    return true
  }
}

/**
 * @template T
 * @param {t.TestCase} tc
 * @param {{users?:number}} conf
 * @param {InitTestObjectCallback<T>} [initTestObject]
 * @return {{testObjects:Array<any>,testConnector:TestConnector,users:Array<TestYInstance>,array0:Y.Array<any>,array1:Y.Array<any>,array2:Y.Array<any>,map0:Y.Map<any>,map1:Y.Map<any>,map2:Y.Map<any>,map3:Y.Map<any>,text0:Y.Text,text1:Y.Text,text2:Y.Text,xml0:Y.XmlElement,xml1:Y.XmlElement,xml2:Y.XmlElement}}
 */
const init = (tc, { users = 5 } = {}, initTestObject) => {
  /**
   * @type {Object<string,any>}
   */
  const result = {
    users: []
  };
  const gen = tc.prng;
  // choose an encoding approach at random
  if (prng__namespace.bool(gen)) {
    useV2Encoding();
  } else {
    useV1Encoding();
  }

  const testConnector = new TestConnector(gen);
  result.testConnector = testConnector;
  for (let i = 0; i < users; i++) {
    const y = testConnector.createY(i);
    y.clientID = i;
    result.users.push(y);
    result['array' + i] = y.getArray('array');
    result['map' + i] = y.getMap('map');
    result['xml' + i] = y.get('xml', Skip.YXmlElement);
    result['text' + i] = y.getText('text');
  }
  testConnector.syncAll();
  result.testObjects = result.users.map(initTestObject || (() => null));
  useV1Encoding();
  return /** @type {any} */ (result)
};

/**
 * @param {Y.IdSet} idSet1
 * @param {Y.IdSet} idSet2
 */
const compareIdSets = (idSet1, idSet2) => {
  t__namespace.assert(idSet1.clients.size === idSet2.clients.size);
  for (const [client, _items1] of idSet1.clients.entries()) {
    const items1 = _items1.getIds();
    const items2 = idSet2.clients.get(client)?.getIds();
    t__namespace.assert(items2 !== undefined && items1.length === items2.length);
    for (let i = 0; i < items1.length; i++) {
      const di1 = items1[i];
      const di2 = /** @type {Array<import('../src/utils/IdSet.js').IdRange>} */ (items2)[i];
      t__namespace.assert(di1.clock === di2.clock && di1.len === di2.len);
    }
  }
  return true
};

/**
 * only use for testing
 *
 * @template T
 * @param {Array<Y.Attribution<T>>} attrs
 * @param {Y.Attribution<T>} attr
 *
 */
const _idmapAttrsHas = (attrs, attr) => {
  const hash = attr.hash();
  return attrs.find(a => a.hash() === hash)
};

/**
 * only use for testing
 *
 * @template T
 * @param {Array<Y.Attribution<T>>} a
 * @param {Array<Y.Attribution<T>>} b
 */
const _idmapAttrsEqual = (a, b) => a.length === b.length && a.every(v => _idmapAttrsHas(b, v));

/**
 * Ensure that all attributes exist. Also create a copy and compare it to the original.
 *
 * @template T
 * @param {Y.IdMap<T>} idmap
 */
const validateIdMap = idmap => {
  const copy = Skip.createIdMap();
  idmap.clients.forEach((ranges, client) => {
    ranges.getIds().forEach(range => {
      range.attrs.forEach(attr => {
        t__namespace.assert(idmap.attrs.has(attr));
        t__namespace.assert(idmap.attrsH.get(attr.hash()) === attr);
        copy.add(client, range.clock, range.len, range.attrs.slice());
      });
    });
    t__namespace.assert(copy.clients.get(client)?.getIds().length === ranges.getIds().length);
  });
  t__namespace.assert(idmap.attrsH.size === idmap.attrs.size);
};

/**
 * @template T
 * @param {Y.IdMap<T>} idmap1
 * @param {Y.IdMap<T>} idmap2
 */
const compareIdmaps = (idmap1, idmap2) => {
  t__namespace.assert(idmap1.clients.size === idmap2.clients.size);
  for (const [client, _items1] of idmap1.clients.entries()) {
    const items1 = _items1.getIds();
    const items2 = idmap2.clients.get(client)?.getIds();
    t__namespace.assert(items2 !== undefined && items1.length === items2.length);
    for (let i = 0; i < items1.length; i++) {
      const di1 = items1[i];
      const di2 = /** @type {Array<import('../src/utils/IdMap.js').AttrRange<T>>} */ (items2)[i];
      t__namespace.assert(di1.clock === di2.clock && di1.len === di2.len && _idmapAttrsEqual(di1.attrs, di2.attrs));
    }
  }
  validateIdMap(idmap1);
  validateIdMap(idmap2);
};

/**
 * @param {prng.PRNG} gen
 * @param {number} clients
 * @param {number} clockRange (max clock - exclusive - by each client)
 */
const createRandomIdSet = (gen, clients, clockRange) => {
  const maxOpLen = 5;
  const numOfOps = math__namespace.ceil((clients * clockRange) / maxOpLen);
  const idset = Skip.createIdSet();
  for (let i = 0; i < numOfOps; i++) {
    const client = prng__namespace.uint32(gen, 0, clients - 1);
    const clockStart = prng__namespace.uint32(gen, 0, clockRange);
    const len = prng__namespace.uint32(gen, 0, clockRange - clockStart);
    Skip.addToIdSet(idset, client, clockStart, len);
  }
  if (idset.clients.size === clients && clients > 1 && prng__namespace.bool(gen)) {
    idset.clients.delete(prng__namespace.uint32(gen, 0, clients));
  }
  return idset
};

/**
 * @template T
 * @param {prng.PRNG} gen
 * @param {number} clients
 * @param {number} clockRange (max clock - exclusive - by each client)
 * @param {Array<T>} attrChoices (max clock - exclusive - by each client)
 * @return {Y.IdMap<T>}
 */
const createRandomIdMap = (gen, clients, clockRange, attrChoices) => {
  const maxOpLen = 5;
  const numOfOps = math__namespace.ceil((clients * clockRange) / maxOpLen);
  const idMap = Skip.createIdMap();
  for (let i = 0; i < numOfOps; i++) {
    const client = prng__namespace.uint32(gen, 0, clients - 1);
    const clockStart = prng__namespace.uint32(gen, 0, clockRange);
    const len = prng__namespace.uint32(gen, 0, clockRange - clockStart);
    const attrs = [prng__namespace.oneOf(gen, attrChoices)];
    // maybe add another attr
    if (prng__namespace.bool(gen)) {
      const a = prng__namespace.oneOf(gen, attrChoices);
      if (attrs.find(attr => attr === a) == null) {
        attrs.push(a);
      }
    }
    idMap.add(client, clockStart, len, attrs.map(v => Skip.createAttributionItem('', v)));
  }
  t__namespace.info(`Created IdMap with ${numOfOps} ranges and ${attrChoices.length} different attributes. Encoded size: ${Skip.encodeIdMap(idMap).byteLength}`);
  return idMap
};

/**
 * 1. reconnect and flush all
 * 2. user 0 gc
 * 3. get type content
 * 4. disconnect & reconnect all (so gc is propagated)
 * 5. compare os, ds, ss
 *
 * @param {Array<TestYInstance>} users
 */
const compare = users => {
  users.forEach(u => u.connect());
  while (users[0].tc.flushAllMessages()) {} // eslint-disable-line
  // For each document, merge all received document updates with Y.mergeUpdates and create a new document which will be added to the list of "users"
  // This ensures that mergeUpdates works correctly
  const mergedDocs = users.map(user => {
    const ydoc = new Skip.Doc();
    exports.enc.applyUpdate(ydoc, exports.enc.mergeUpdates(user.updates));
    return ydoc
  });
  users.push(.../** @type {any} */(mergedDocs));
  const userArrayValues = users.map(u => u.getArray('array').toJSON());
  const userMapValues = users.map(u => u.getMap('map').toJSON());
  // @todo fix type error here
  // @ts-ignore
  const userXmlValues = users.map(u => /** @type {Y.XmlElement} */ (u.get('xml', Skip.YXmlElement)).toString());
  const userTextValues = users.map(u => u.getText('text').getContentDeep());
  for (const u of users) {
    t__namespace.assert(u.store.pendingDs === null);
    t__namespace.assert(u.store.pendingStructs === null);
  }
  // Test Array iterator
  t__namespace.compare(users[0].getArray('array').toArray(), Array.from(users[0].getArray('array')));
  // Test Map iterator
  const ymapkeys = Array.from(users[0].getMap('map').keys());
  t__namespace.assert(ymapkeys.length === Object.keys(userMapValues[0]).length);
  ymapkeys.forEach(key => t__namespace.assert(object__namespace.hasProperty(userMapValues[0], key)));
  /**
   * @type {Object<string,any>}
   */
  const mapRes = {};
  for (const [k, v] of users[0].getMap('map')) {
    mapRes[k] = v instanceof Skip.AbstractType ? v.toJSON() : v;
  }
  t__namespace.compare(userMapValues[0], mapRes);
  // Compare all users
  for (let i = 0; i < users.length - 1; i++) {
    t__namespace.compare(userArrayValues[i].length, users[i].getArray('array').length);
    t__namespace.compare(userArrayValues[i], userArrayValues[i + 1]);
    t__namespace.compare(userMapValues[i], userMapValues[i + 1]);
    t__namespace.compare(userXmlValues[i], userXmlValues[i + 1]);
    t__namespace.compare(list__namespace.toArray(userTextValues[i].children).map(a => (delta__namespace.$textOp.check(a) || delta__namespace.$insertOp.check(a)) ? a.insert.length : 0).reduce((a, b) => a + b, 0), users[i].getText('text').length);
    t__namespace.compare(userTextValues[i], userTextValues[i + 1], '', (_constructor, a, b) => {
      if (a instanceof Skip.AbstractType) {
        t__namespace.compare(a.toJSON(), b.toJSON());
      } else if (a !== b) {
        t__namespace.fail('Deltas dont match');
      }
      return true
    });
    t__namespace.compare(Skip.encodeStateVector(users[i]), Skip.encodeStateVector(users[i + 1]));
    Skip.equalIdSets(Skip.createDeleteSetFromStructStore(users[i].store), Skip.createDeleteSetFromStructStore(users[i + 1].store));
    compareStructStores(users[i].store, users[i + 1].store);
    t__namespace.compare(Skip.encodeSnapshot(Skip.snapshot(users[i])), Skip.encodeSnapshot(Skip.snapshot(users[i + 1])));
  }
  users.forEach(user => {
    compareIdSets(user.store.ds, Skip.createDeleteSetFromStructStore(user.store));
  });
  users.map(u => u.destroy());
};

/**
 * @param {Y.Item?} a
 * @param {Y.Item?} b
 * @return {boolean}
 */
const compareItemIDs = (a, b) => a === b || (a !== null && b != null && Skip.compareIDs(a.id, b.id));

/**
 * @param {import('../src/internals.js').StructStore} ss1
 * @param {import('../src/internals.js').StructStore} ss2
 */
const compareStructStores = (ss1, ss2) => {
  t__namespace.assert(ss1.clients.size === ss2.clients.size);
  for (const [client, structs1] of ss1.clients) {
    const structs2 = /** @type {Array<Y.AbstractStruct>} */ (ss2.clients.get(client));
    t__namespace.assert(structs2 !== undefined && structs1.length === structs2.length);
    for (let i = 0; i < structs1.length; i++) {
      const s1 = structs1[i];
      const s2 = structs2[i];
      // checks for abstract struct
      if (
        s1.constructor !== s2.constructor ||
        !Skip.compareIDs(s1.id, s2.id) ||
        s1.deleted !== s2.deleted ||
        // @ts-ignore
        s1.length !== s2.length
      ) {
        t__namespace.fail('Structs dont match');
      }
      if (s1 instanceof Skip.Item) {
        if (
          !(s2 instanceof Skip.Item) ||
          !((s1.left === null && s2.left === null) || (s1.left !== null && s2.left !== null && Skip.compareIDs(s1.left.lastId, s2.left.lastId))) ||
          !compareItemIDs(s1.right, s2.right) ||
          !Skip.compareIDs(s1.origin, s2.origin) ||
          !Skip.compareIDs(s1.rightOrigin, s2.rightOrigin) ||
          s1.parentSub !== s2.parentSub
        ) {
          return t__namespace.fail('Items dont match')
        }
        // make sure that items are connected correctly
        t__namespace.assert(s1.left === null || s1.left.right === s1);
        t__namespace.assert(s1.right === null || s1.right.left === s1);
        t__namespace.assert(s2.left === null || s2.left.right === s2);
        t__namespace.assert(s2.right === null || s2.right.left === s2);
      }
    }
  }
};

/**
 * @template T
 * @callback InitTestObjectCallback
 * @param {TestYInstance} y
 * @return {T}
 */

/**
 * @template T
 * @param {t.TestCase} tc
 * @param {Array<function(Y.Doc,prng.PRNG,T):void>} mods
 * @param {number} iterations
 * @param {InitTestObjectCallback<T>} [initTestObject]
 */
const applyRandomTests = (tc, mods, iterations, initTestObject) => {
  const gen = tc.prng;
  const result = init(tc, { users: 5 }, initTestObject);
  const { testConnector, users } = result;
  for (let i = 0; i < iterations; i++) {
    if (prng__namespace.int32(gen, 0, 100) <= 2) {
      // 2% chance to disconnect/reconnect a random user
      if (prng__namespace.bool(gen)) {
        testConnector.disconnectRandom();
      } else {
        testConnector.reconnectRandom();
      }
    } else if (prng__namespace.int32(gen, 0, 100) <= 1) {
      // 1% chance to flush all
      testConnector.flushAllMessages();
    } else if (prng__namespace.int32(gen, 0, 100) <= 50) {
      // 50% chance to flush a random message
      testConnector.flushRandomMessage();
    }
    const user = prng__namespace.int32(gen, 0, users.length - 1);
    const test = prng__namespace.oneOf(gen, mods);
    test(users[user], gen, result.testObjects[user]);
  }
  compare(users);
  return result
};

exports.AbsolutePosition = Skip.AbsolutePosition;
exports.AbstractAttributionManager = Skip.AbstractAttributionManager;
exports.AbstractConnector = Skip.AbstractConnector;
exports.AbstractStruct = Skip.AbstractStruct;
exports.AbstractType = Skip.AbstractType;
exports.Array = Skip.YArray;
exports.Attribution = Skip.AttributionItem;
exports.ContentAny = Skip.ContentAny;
exports.ContentBinary = Skip.ContentBinary;
exports.ContentDeleted = Skip.ContentDeleted;
exports.ContentDoc = Skip.ContentDoc;
exports.ContentEmbed = Skip.ContentEmbed;
exports.ContentFormat = Skip.ContentFormat;
exports.ContentJSON = Skip.ContentJSON;
exports.ContentString = Skip.ContentString;
exports.ContentType = Skip.ContentType;
exports.DiffAttributionManager = Skip.DiffAttributionManager;
exports.Doc = Skip.Doc;
exports.GC = Skip.GC;
exports.ID = Skip.ID;
exports.IdMap = Skip.IdMap;
exports.IdSet = Skip.IdSet;
exports.Item = Skip.Item;
exports.Map = Skip.YMap;
exports.RelativePosition = Skip.RelativePosition;
exports.Skip = Skip.Skip;
exports.Snapshot = Skip.Snapshot;
exports.Text = Skip.YText;
exports.Transaction = Skip.Transaction;
exports.TwosetAttributionManager = Skip.TwosetAttributionManager;
exports.UndoManager = Skip.UndoManager;
exports.UpdateDecoderV1 = Skip.UpdateDecoderV1;
exports.UpdateDecoderV2 = Skip.UpdateDecoderV2;
exports.UpdateEncoderV1 = Skip.UpdateEncoderV1;
exports.UpdateEncoderV2 = Skip.UpdateEncoderV2;
exports.XmlElement = Skip.YXmlElement;
exports.XmlFragment = Skip.YXmlFragment;
exports.XmlHook = Skip.YXmlHook;
exports.XmlText = Skip.YXmlText;
exports.YEvent = Skip.YEvent;
exports.applyUpdate = Skip.applyUpdate;
exports.applyUpdateV2 = Skip.applyUpdateV2;
exports.cleanupYTextFormatting = Skip.cleanupYTextFormatting;
exports.cloneDoc = Skip.cloneDoc;
exports.compareIDs = Skip.compareIDs;
exports.compareRelativePositions = Skip.compareRelativePositions;
exports.convertUpdateFormatV1ToV2 = Skip.convertUpdateFormatV1ToV2;
exports.convertUpdateFormatV2ToV1 = Skip.convertUpdateFormatV2ToV1;
exports.createAbsolutePositionFromRelativePosition = Skip.createAbsolutePositionFromRelativePosition;
exports.createAttributionItem = Skip.createAttributionItem;
exports.createAttributionManagerFromDiff = Skip.createAttributionManagerFromDiff;
exports.createDeleteSetFromStructStore = Skip.createDeleteSetFromStructStore;
exports.createDocFromSnapshot = Skip.createDocFromSnapshot;
exports.createID = Skip.createID;
exports.createIdMap = Skip.createIdMap;
exports.createIdMapFromIdSet = Skip.createIdMapFromIdSet;
exports.createIdSet = Skip.createIdSet;
exports.createInsertionSetFromStructStore = Skip.createInsertSetFromStructStore;
exports.createRelativePositionFromJSON = Skip.createRelativePositionFromJSON;
exports.createRelativePositionFromTypeIndex = Skip.createRelativePositionFromTypeIndex;
exports.createSnapshot = Skip.createSnapshot;
exports.decodeIdMap = Skip.decodeIdMap;
exports.decodeRelativePosition = Skip.decodeRelativePosition;
exports.decodeSnapshot = Skip.decodeSnapshot;
exports.decodeSnapshotV2 = Skip.decodeSnapshotV2;
exports.decodeStateVector = Skip.decodeStateVector;
exports.decodeUpdate = Skip.decodeUpdate;
exports.decodeUpdateV2 = Skip.decodeUpdateV2;
exports.diffDocsToDelta = Skip.diffDocsToDelta;
exports.diffIdMap = Skip.diffIdMap;
exports.diffIdSet = Skip.diffIdSet;
exports.diffUpdate = Skip.diffUpdate;
exports.diffUpdateV2 = Skip.diffUpdateV2;
exports.emptySnapshot = Skip.emptySnapshot;
exports.encodeIdMap = Skip.encodeIdMap;
exports.encodeRelativePosition = Skip.encodeRelativePosition;
exports.encodeSnapshot = Skip.encodeSnapshot;
exports.encodeSnapshotV2 = Skip.encodeSnapshotV2;
exports.encodeStateAsUpdate = Skip.encodeStateAsUpdate;
exports.encodeStateAsUpdateV2 = Skip.encodeStateAsUpdateV2;
exports.encodeStateVector = Skip.encodeStateVector;
exports.encodeStateVectorFromUpdate = Skip.encodeStateVectorFromUpdate;
exports.encodeStateVectorFromUpdateV2 = Skip.encodeStateVectorFromUpdateV2;
exports.equalIdSets = Skip.equalIdSets;
exports.equalSnapshots = Skip.equalSnapshots;
exports.findIndexSS = Skip.findIndexSS;
exports.findRootTypeKey = Skip.findRootTypeKey;
exports.getItem = Skip.getItem;
exports.getItemCleanEnd = Skip.getItemCleanEnd;
exports.getItemCleanStart = Skip.getItemCleanStart;
exports.getState = Skip.getState;
exports.getTypeChildren = Skip.getTypeChildren;
exports.insertIntoIdMap = Skip.insertIntoIdMap;
exports.insertIntoIdSet = Skip.insertIntoIdSet;
exports.isParentOf = Skip.isParentOf;
exports.iterateStructsByIdSet = Skip.iterateStructsByIdSet;
exports.logType = Skip.logType;
exports.logUpdate = Skip.logUpdate;
exports.logUpdateV2 = Skip.logUpdateV2;
exports.mergeIdMaps = Skip.mergeIdMaps;
exports.mergeIdSets = Skip.mergeIdSets;
exports.mergeUpdates = Skip.mergeUpdates;
exports.mergeUpdatesV2 = Skip.mergeUpdatesV2;
exports.noAttributionsManager = Skip.noAttributionsManager;
exports.obfuscateUpdate = Skip.obfuscateUpdate;
exports.obfuscateUpdateV2 = Skip.obfuscateUpdateV2;
exports.readIdMap = Skip.readIdMap;
exports.readIdSet = Skip.readIdSet;
exports.readUpdate = Skip.readUpdate;
exports.readUpdateIdRanges = Skip.readUpdateIdRanges;
exports.readUpdateIdRangesV2 = Skip.readUpdateIdRangesV2;
exports.readUpdateV2 = Skip.readUpdateV2;
exports.relativePositionToJSON = Skip.relativePositionToJSON;
exports.snapshot = Skip.snapshot;
exports.snapshotContainsUpdate = Skip.snapshotContainsUpdate;
exports.transact = Skip.transact;
exports.tryGc = Skip.tryGc;
exports.typeListToArraySnapshot = Skip.typeListToArraySnapshot;
exports.typeMapGetAllSnapshot = Skip.typeMapGetAllSnapshot;
exports.typeMapGetSnapshot = Skip.typeMapGetSnapshot;
exports.TestConnector = TestConnector;
exports.TestYInstance = TestYInstance;
exports._idmapAttrsEqual = _idmapAttrsEqual;
exports.applyRandomTests = applyRandomTests;
exports.compare = compare;
exports.compareIdSets = compareIdSets;
exports.compareIdmaps = compareIdmaps;
exports.compareItemIDs = compareItemIDs;
exports.compareStructStores = compareStructStores;
exports.createRandomIdMap = createRandomIdMap;
exports.createRandomIdSet = createRandomIdSet;
exports.encV1 = encV1;
exports.encV2 = encV2;
exports.init = init;
exports.validateIdMap = validateIdMap;
