// @y/y/dist/yjs.cjs
'use strict';

var Skip = require('./Skip-j0kX7pdq.js');
require('lib0/function');
require('lib0/decoding');
require('lib0/binary');
require('lib0/map');
require('./index-R7GxO-36.js');
require('lib0/observable');
require('lib0/array');
require('lib0/math');
require('lib0/encoding');
require('lib0/traits');
require('lib0/random');
require('lib0/promise');
require('lib0/buffer');
require('lib0/error');
require('lib0/set');
require('lib0/logging');
require('lib0/time');
require('lib0/string');
require('lib0/delta');
require('lib0/hash/rabin');
require('lib0/iterator');
require('lib0/object');
require('lib0/environment');
require('lib0/schema');



exports.AbsolutePosition = Skip.AbsolutePosition;
exports.AbstractAttributionManager = Skip.AbstractAttributionManager;
exports.AbstractConnector = Skip.AbstractConnector;
exports.AbstractStruct = Skip.AbstractStruct;
exports.AbstractType = Skip.AbstractType;
exports.Array = Skip.YArray;
exports.Attribution = Skip.AttributionItem;
exports.ContentAny = Skip.ContentAny;
exports.ContentBinary = Skip.ContentBinary;
exports.ContentDeleted = Skip.ContentDeleted;
exports.ContentDoc = Skip.ContentDoc;
exports.ContentEmbed = Skip.ContentEmbed;
exports.ContentFormat = Skip.ContentFormat;
exports.ContentJSON = Skip.ContentJSON;
exports.ContentString = Skip.ContentString;
exports.ContentType = Skip.ContentType;
exports.DiffAttributionManager = Skip.DiffAttributionManager;
exports.Doc = Skip.Doc;
exports.GC = Skip.GC;
exports.ID = Skip.ID;
exports.IdMap = Skip.IdMap;
exports.IdSet = Skip.IdSet;
exports.Item = Skip.Item;
exports.Map = Skip.YMap;
exports.RelativePosition = Skip.RelativePosition;
exports.Skip = Skip.Skip;
exports.Snapshot = Skip.Snapshot;
exports.Text = Skip.YText;
exports.Transaction = Skip.Transaction;
exports.TwosetAttributionManager = Skip.TwosetAttributionManager;
exports.UndoManager = Skip.UndoManager;
exports.UpdateDecoderV1 = Skip.UpdateDecoderV1;
exports.UpdateDecoderV2 = Skip.UpdateDecoderV2;
exports.UpdateEncoderV1 = Skip.UpdateEncoderV1;
exports.UpdateEncoderV2 = Skip.UpdateEncoderV2;
exports.XmlElement = Skip.YXmlElement;
exports.XmlFragment = Skip.YXmlFragment;
exports.XmlHook = Skip.YXmlHook;
exports.XmlText = Skip.YXmlText;
exports.YEvent = Skip.YEvent;
exports.applyUpdate = Skip.applyUpdate;
exports.applyUpdateV2 = Skip.applyUpdateV2;
exports.cleanupYTextFormatting = Skip.cleanupYTextFormatting;
exports.cloneDoc = Skip.cloneDoc;
exports.compareIDs = Skip.compareIDs;
exports.compareRelativePositions = Skip.compareRelativePositions;
exports.convertUpdateFormatV1ToV2 = Skip.convertUpdateFormatV1ToV2;
exports.convertUpdateFormatV2ToV1 = Skip.convertUpdateFormatV2ToV1;
exports.createAbsolutePositionFromRelativePosition = Skip.createAbsolutePositionFromRelativePosition;
exports.createAttributionItem = Skip.createAttributionItem;
exports.createAttributionManagerFromDiff = Skip.createAttributionManagerFromDiff;
exports.createDeleteSetFromStructStore = Skip.createDeleteSetFromStructStore;
exports.createDocFromSnapshot = Skip.createDocFromSnapshot;
exports.createID = Skip.createID;
exports.createIdMap = Skip.createIdMap;
exports.createIdMapFromIdSet = Skip.createIdMapFromIdSet;
exports.createIdSet = Skip.createIdSet;
exports.createInsertionSetFromStructStore = Skip.createInsertSetFromStructStore;
exports.createRelativePositionFromJSON = Skip.createRelativePositionFromJSON;
exports.createRelativePositionFromTypeIndex = Skip.createRelativePositionFromTypeIndex;
exports.createSnapshot = Skip.createSnapshot;
exports.decodeIdMap = Skip.decodeIdMap;
exports.decodeRelativePosition = Skip.decodeRelativePosition;
exports.decodeSnapshot = Skip.decodeSnapshot;
exports.decodeSnapshotV2 = Skip.decodeSnapshotV2;
exports.decodeStateVector = Skip.decodeStateVector;
exports.decodeUpdate = Skip.decodeUpdate;
exports.decodeUpdateV2 = Skip.decodeUpdateV2;
exports.diffDocsToDelta = Skip.diffDocsToDelta;
exports.diffIdMap = Skip.diffIdMap;
exports.diffIdSet = Skip.diffIdSet;
exports.diffUpdate = Skip.diffUpdate;
exports.diffUpdateV2 = Skip.diffUpdateV2;
exports.emptySnapshot = Skip.emptySnapshot;
exports.encodeIdMap = Skip.encodeIdMap;
exports.encodeRelativePosition = Skip.encodeRelativePosition;
exports.encodeSnapshot = Skip.encodeSnapshot;
exports.encodeSnapshotV2 = Skip.encodeSnapshotV2;
exports.encodeStateAsUpdate = Skip.encodeStateAsUpdate;
exports.encodeStateAsUpdateV2 = Skip.encodeStateAsUpdateV2;
exports.encodeStateVector = Skip.encodeStateVector;
exports.encodeStateVectorFromUpdate = Skip.encodeStateVectorFromUpdate;
exports.encodeStateVectorFromUpdateV2 = Skip.encodeStateVectorFromUpdateV2;
exports.equalIdSets = Skip.equalIdSets;
exports.equalSnapshots = Skip.equalSnapshots;
exports.findIndexSS = Skip.findIndexSS;
exports.findRootTypeKey = Skip.findRootTypeKey;
exports.getItem = Skip.getItem;
exports.getItemCleanEnd = Skip.getItemCleanEnd;
exports.getItemCleanStart = Skip.getItemCleanStart;
exports.getState = Skip.getState;
exports.getTypeChildren = Skip.getTypeChildren;
exports.insertIntoIdMap = Skip.insertIntoIdMap;
exports.insertIntoIdSet = Skip.insertIntoIdSet;
exports.isParentOf = Skip.isParentOf;
exports.iterateStructsByIdSet = Skip.iterateStructsByIdSet;
exports.logType = Skip.logType;
exports.logUpdate = Skip.logUpdate;
exports.logUpdateV2 = Skip.logUpdateV2;
exports.mergeIdMaps = Skip.mergeIdMaps;
exports.mergeIdSets = Skip.mergeIdSets;
exports.mergeUpdates = Skip.mergeUpdates;
exports.mergeUpdatesV2 = Skip.mergeUpdatesV2;
exports.noAttributionsManager = Skip.noAttributionsManager;
exports.obfuscateUpdate = Skip.obfuscateUpdate;
exports.obfuscateUpdateV2 = Skip.obfuscateUpdateV2;
exports.readIdMap = Skip.readIdMap;
exports.readIdSet = Skip.readIdSet;
exports.readUpdate = Skip.readUpdate;
exports.readUpdateIdRanges = Skip.readUpdateIdRanges;
exports.readUpdateIdRangesV2 = Skip.readUpdateIdRangesV2;
exports.readUpdateV2 = Skip.readUpdateV2;
exports.relativePositionToJSON = Skip.relativePositionToJSON;
exports.snapshot = Skip.snapshot;
exports.snapshotContainsUpdate = Skip.snapshotContainsUpdate;
exports.transact = Skip.transact;
exports.tryGc = Skip.tryGc;
exports.typeListToArraySnapshot = Skip.typeListToArraySnapshot;
exports.typeMapGetAllSnapshot = Skip.typeMapGetAllSnapshot;
exports.typeMapGetSnapshot = Skip.typeMapGetSnapshot;
